Gauge Contracts

JakeHartnell opened the pull request:
Inspired by the [Curve gauge system](https://resources.curve.fi/reward-gauges/understanding-gauges/) and forked from the [Wynd DAO repo](https://github.com/wynddao/wynddao) with history and Apache 2.0 license maintained.

These contracts have been modified to be more generic and support any type of DAO.

However, consider this a straw-man implementation. There are many improvements to be made!
- [ ] More tests (consider refactoring with cw-orch)
- [ ] Make improvements to `GaugeAdapter` API
- [ ] Improve `GaugeAdapter` example (needs significant work, we may want more than one example)
- [ ] Currently oes not handle small voting power numbers elegantly (a problem with NFT DAOs, where people might stake 1 NFT)
- [ ] Better READMEs and documentation

Bonus:
- [ ] Implement decay, so that people eventually have to re-express their preferences on the gauge
- [ ] Think about how to hook in incentives from the `dao-rewards-distributor` contract so that DAOs can optional incentivize people to participate in gauges

If you are looking to contribute to this and want to understand how they work, the `gauge-orchestrator` [README](https://github.com/DA0-DA0/dao-contracts/tree/gauges-are-cool/contracts/gauges/gauge) is a good place to start.

[Twitter post and discussion.](https://x.com/JakeHartnell/status/1809247777567748187)

Gauges in actual decentralize applications:
- [Curve](https://dao.curve.fi/gaugeweight): the OG
- [Wynd](https://app.wynddao.com/gauges): what these are based on

Use cases:
- Budget Allocation
- Setting parameters
- Incentives Allocation
- Managing liquid staking protocols (choosing which validators to to delegate to)
- etc.

ueco-jb committed:
Update to v1.2.1

dybbuk-light committed:
Release 1.4.0

ueco-jb committed:
Update contracts to version 1.5.4

ueco-jb committed:
Set version v1.6.0

ueco-jb committed:
Merge pull request #8 from cosmorama/release-v1-6-0

Set version v1.6.0

ueco-jb committed:
Set version: 1.7.0

ueco-jb committed:
Merge pull request #10 from wynddao/release-v-1-7-0

Set version: 1.7.0

Commit:
gauge tests pass

Commit:
Support all DAO types for gauges

Commit:
Merge branch 'gauges-for-daos' into gauges-are-cool

Commit:
Make gauges folder

Commit:
Add support for all DAO types

Commit:
Clean up, appease clippy gods

hard-nett commented:
> * [ ]  More tests (consider refactoring with cw-orch)
> * [ ]  Make improvements to `GaugeAdapter` API
> * [ ]  Improve `GaugeAdapter` example (needs significant work, we may want more than one example)

spearheading these three points this week! an additional gauge adapter template can be one that turns itself off after a specific # of times the gauges epoch occurs.

JakeHartnell commented:
> > * [ ]   More tests (consider refactoring with cw-orch)
> > * [ ]   Make improvements to `GaugeAdapter` API
> > * [ ]   Improve `GaugeAdapter` example (needs significant work, we may want more than one example)
> 
> spearheading these three points this week! an additional gauge adapter template can be one that turns itself off after a specific # of times the gauges epoch occurs.

AMAZING :heart:

ismellike committed:
Use cw-denom and payment utils in gauge adapter (#847)

* Use cw-denom and payment utils in gauge adapter

TODO: 2 tests are failing, because the cw20 cannot be validated with cw-denom. This should be fixed with a cw-orch refactor.

* gauge adapter cw-orch tests

* include suite changes

* cleanup adapter test

* A couple tweaks to gauge adapter tests

* cleanup tests a bit

* cargo lock

* bump prost, lint

* lint

---------

Co-authored-by: hard-nett <hardnettt@proton.me>
Co-authored-by: Hard-Nett <123711748+hard-nett@users.noreply.github.com>

JakeHartnell commented on contracts/gauges/gauge/src/msg.rs:
```diff
@@ -0,0 +1,263 @@
+pub struct InstantiateMsg {
+    /// Address of contract to that contains all voting powers (where we query)
+    pub voting_powers: String,
+    /// Addres that will call voting power change hooks (often same as voting power contract)
```
```suggestion
    /// Address that will call voting power change hooks (often same as voting power contract)
```

Commit:
Merged latest development

Commit:
debug

Commit:
no
